qemu/tests/numa_cpu.py

```diff
@@ -0,0 +1,149 @@
+        """
+        Convert the cpu topology to cpu index list
+        """
+        def _get_boundary(value, max_value, weight):
+            """
+            Get the data range of one bit
+
+            :param value: the current value of the bit
+            :param max_value: the max value of the bit
+            :param weight: the weight of the bit
+            """
+            min_boundary = int(value if value is not None else 0) * weight
+            max_boundary = int(value if value is not None else (max_value - 1)) * weight
+            return (min_boundary, max_boundary)
+
+        if vm_arch in ('x86_64', 'i386'):
+            socket_min, socket_max = _get_boundary(socket, vcpu_sockets, socket_weight)
+            die_min, die_max = _get_boundary(die, vcpu_dies, die_weight)
+            core_min, core_max = _get_boundary(core, vcpu_cores, core_weight)
+            thread_min, thread_max = _get_boundary(thread, vcpu_threads, thread_weight)
+            cpu_min = socket_min + die_min + core_min + thread_min
+            cpu_max = socket_max + die_max + core_max + thread_max
+        elif vm_arch in ('ppc64', 'ppc64le'):
+            cpu_min = int(core)
+            cpu_max = int(core) + vcpu_threads - 1
+        cpu_index = list(range(cpu_min, cpu_max + 1))
+        return cpu_index
+
+    def numa_cpu_guest():
+        """
+        Get the numa cpus for each node in guest os, linux only
+        """
+        error_context.context("Get cpus in guest os", logging.info)
+        numa_info_guest = NumaInfo(session=session)
+        nodes_guest = numa_info_guest.online_nodes
+        numa_cpu_guest = []
+        for node in nodes_guest:
+            numa_cpus = numa_info_guest.online_nodes_cpus[node]
+            numa_cpus = set([int(v) for v in numa_cpus.split()])
+            numa_cpu_guest.append(numa_cpus)
+        return numa_cpu_guest
+
+    def numa_cpu_cli():
+        """
+        Get the numa cpus for each node according to the qemu cli
+        """
+        error_context.context("Get the expected cpus in qemu command line", logging.info)
+        numa_cpus = params.objects("guest_numa_cpus")
+        numa_cpu_cli = []
+        tmp = {}
+        for numa_cpu in numa_cpus:
+            numa_cpu_params = params.object_params(numa_cpu)
+            nodeid = numa_cpu_params["numa_cpu_nodeid"]
+            socket = numa_cpu_params.get("numa_cpu_socketid")
+            die = numa_cpu_params.get("numa_cpu_dieid")
+            core = numa_cpu_params.get("numa_cpu_coreid")
+            thread = numa_cpu_params.get("numa_cpu_threadid")
+            cpu_list = convert_cpu_topology_to_index(socket, die, core, thread)
+            if nodeid in tmp.keys():
+                tmp[nodeid] += cpu_list
+            else:
+                tmp[nodeid] = cpu_list
+        for item in sorted(tmp.items(), key=lambda item: item[0]):
+            numa_cpu_cli.append(set(sorted(item[1])))
+        return numa_cpu_cli
```

Comment:
How about save numa_cpu_cli as a dict just like specified_cpus and numa_cpu_monitor? So you can compare the dict directly in L139.

Reply:
compare the whole dict directly ?  Wouldn't it be better to show user which node test failed ?

Reply:
By the way, `specified_cpus` is a list instead of dict, use the list index as nodenr.

Reply:
Yeah, I see. All of there variants store the cpus info for every numa node, I guess we can use one same data structure.
Current comparison only show the first mismatched node. A whole info of all nodes would be more helpful to debug.

Reply:
> Yeah, I see. All of there variants store the cpus info for every numa node, I guess we can use one same data structure.
> Current comparison only show the first mismatched node. A whole info of all nodes would be more helpful to debug.

Sure, all of the whole info for cli, monitor, guest os  already be shown in debug.log after get it.
and the data structure for them are all same now.

Reply:
@huangyum   Do you have any more comments ?